tools/testrunner/log.go

```diff
@@ -65,3 +66,285 @@ func testOnlyStacktrace(stacktrace string) string {
+		name = name[:p]
+	}
+	return name, true
+}
+
+// tryParseDataRace parses a data race alert at position i if present.
+func tryParseDataRace(lines []string, i int, line string) (alert, int, bool) {
+	if !strings.HasPrefix(line, "WARNING: DATA RACE") {
+		return alert{}, i, false
+	}
+	start := findRaceBlockStart(lines, i)
+	block, end := collectBlock(lines, start, shouldStopDataRace)
+	return alert{
+		Kind:    alertKindDataRace,
+		Summary: "Data race detected",
+		Details: block,
+		Tests:   extractTestNames(block),
+	}, end, true
+}
+
+// tryParsePanic parses a non-timeout panic alert at position i if present.
+func tryParsePanic(lines []string, i int, line string) (alert, int, bool) {
+	if !strings.HasPrefix(line, "panic: ") || strings.HasPrefix(line, "panic: test timed out after") {
+		return alert{}, i, false
+	}
+	block, end := collectBlock(lines, i, shouldStopOnTestBoundary)
+	return alert{
+		Kind:    alertKindPanic,
+		Summary: strings.TrimSpace(strings.TrimPrefix(line, "panic: ")),
+		Details: block,
+		Tests:   extractTestNames(block),
+	}, end, true
+}
+
+// tryParseFatal parses a runtime fatal error alert at position i if present.
+func tryParseFatal(lines []string, i int, line string) (alert, int, bool) {
+	if !strings.HasPrefix(line, "fatal error: ") {
+		return alert{}, i, false
+	}
+	block, end := collectBlock(lines, i, shouldStopOnTestBoundary)
+	return alert{
+		Kind:    alertKindFatal,
+		Summary: strings.TrimSpace(strings.TrimPrefix(line, "fatal error: ")),
+		Details: block,
+		Tests:   extractTestNames(block),
+	}, end, true
+}
+
+// findRaceBlockStart searches upward for the race report delimiter.
+func findRaceBlockStart(lines []string, i int) int {
+	start := i
+	for j := i - 1; j >= 0; j-- {
+		if isRaceBoundary(lines[j]) {
+			start = j
+			break
+		}
+	}
+	return start
+}
+
+// collectBlock builds a block from start until the stop condition is met.
+func collectBlock(lines []string, start int, stop func(line string, idx, start int) bool) (string, int) {
+	var b strings.Builder
+	end := len(lines) - 1
+	for j := start; j < len(lines); j++ {
+		b.WriteString(lines[j])
+		b.WriteByte('\n')
+		if stop(lines[j], j, start) {
+			end = j
+			break
+		}
+		end = j
+	}
+	return b.String(), end
+}
+
+func isRaceBoundary(line string) bool {
+	return strings.HasPrefix(strings.TrimSpace(line), "==================")
+}
+
+func isTestResultBoundary(line string) bool {
+	return strings.HasPrefix(line, "FAIL") || strings.HasPrefix(line, "PASS")
+}
+
+func shouldStopDataRace(line string, idx, start int) bool {
+	if idx > start && isRaceBoundary(line) {
+		return true
+	}
+	return isTestResultBoundary(line)
+}
+
+func shouldStopOnTestBoundary(line string, _ int, _ int) bool {
+	return isTestResultBoundary(line)
+}
+
+// dedupeAlerts removes duplicate alerts (e.g., repeated across retries) based
+// on kind and details while preserving the first-seen order.
+func dedupeAlerts(alerts []alert) []alert {
```

Comment:
Would it make sense to do this in the JUnit merge part? That way it would be alert-agnostic and seems like a good place for this functionality.